core/include/detray/tools/bin_fillers.hpp

```diff
@@ -0,0 +1,139 @@
+    };
+
+    template <typename detector_t, typename volume_type, typename grid_t>
+    DETRAY_HOST auto operator()(grid_t &grid, const detector_t &det,
+                                const volume_type &vol,
+                                const typename detector_t::geometry_context ctx,
+                                std::vector<bin_data<grid_t>> &bins) const
+        -> void {
+        this->operator()(grid, detray::ranges::subrange(det.surfaces(), vol),
+                         det.transform_store(), det.mask_store(), ctx, bins);
+    }
+
+    template <typename grid_t, typename surface_container,
+              typename mask_container, typename transform_container>
+    DETRAY_HOST auto operator()(
+        grid_t &grid, const surface_container &, const transform_container &,
+        const mask_container &,
+        const typename transform_container::context_type,
+        std::vector<bin_data<grid_t>> &bins) const -> void {
+        for (const bin_data<grid_t> &bd : bins) {
+            grid.populate(bd.local_bin_idx, bd.single_element);
+        }
+    }
+};
+
+/// Fill a surface grid using the surface translation.
+///
+/// @param grid the grid that should be filled
+/// @param det the detector from which to get the surface placements
+/// @param vol the volume the grid belongs to
+/// @param ctx the geometry context
+struct fill_by_pos {
+
+    template <typename detector_t, typename volume_type, typename grid_t,
+              typename... Args>
+    DETRAY_HOST auto operator()(grid_t &grid, const detector_t &det,
+                                const volume_type &vol,
+                                const typename detector_t::geometry_context ctx,
+                                Args &&...) const -> void {
+        this->operator()(grid, detray::ranges::subrange(det.surfaces(), vol),
+                         det.transform_store(), det.mask_store(), ctx);
+    }
+
+    template <typename grid_t, typename surface_container,
+              typename mask_container, typename transform_container,
+              typename... Args>
+    DETRAY_HOST auto operator()(
+        grid_t &grid, const surface_container &surfaces,
+        const transform_container &transforms, const mask_container & /*masks*/,
+        const typename transform_container::context_type ctx, Args &&...) const
+        -> void {
+
+        // Fill the volumes surfaces into the grid
+        for (const auto &[idx, sf] : detray::views::enumerate(surfaces)) {
+            // TODO: Remove after toy geo is switched to new builders
+            if (sf.is_portal()) {
+                continue;
+            }
+            // no partals in grids allowed
+            assert(not sf.is_portal());
+
+            const auto &sf_trf = transforms.at(sf.transform(), ctx);
+            const auto &t = sf_trf.translation();
+            const auto loc_pos = grid.global_to_local(
+                __plugin::transform3<detray::scalar>{}, t, t);
```

Comment:
I think you can use `transform3_type` from `detector_t`

Reply:
good point